### ISSUE TITLE
Creating a Outcome Diff

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/action#182]
Let's take a simple project that has a 3 changes in its life.
It starts out being owned by Jordan. He adds a description and passed it off to me. Then, I complete it:

```
const buildWebapp0 = {
  id: 'team123::999999',
  content: null,
  type: 'project',
  status: 'active',
  teamMemberId: 'userJordan::team123'
};

const buildWebapp1 = {
  id: 'team123::999999',
  content: 'Build a realtime app',
  type: 'project',
  status: 'active',
  teamMemberId: 'userMatt::team123'
};

const buildWebapp2 = {
  id: 'team123::999999',
  content: 'Build a realtime app',
  type: 'project',
  status: 'done',
  teamMemberId: 'userMatt::team123'
};
```

What should the diff log look like? It's important to get this right because if we do, we can get a LOT of information from this.

We have some options:
- We could store every field at every step, so it looks like it does above
- We could store just the fields that change from the previous step
- We could store an `updatedAt` field and index on it so we can replay the items
- We could store a `meetingId` field on it if the change occurred during a meeting. (If we don't do this, we could still get the meeting ID by matching the `updatedAt` fields to the `Meeting.endedAt` field.
 
There are a few considerations:
- What if someone clicks 'stuck' by accident & quickly changes it back to 'active'? Is this worth storing? Maybe yes, but during the replay we can see that the duration was < 10 seconds and remove it from reporting?
- In the meeting summary / summary email, do we need the previous & current state, or will current suffice? For example, let's say we changed the title from "build an app" to " build a realtime webapp" and changed the status from active to stuck. In our summary, is it enough to say: "For the project: 'build a realtime webapp" we changed the title, and set the status to stuck". This point is nontrivial because if we save just the fields that change, to get the previous state, we'd need to replay all the changes starting with the earliest.

**Comment by mattkrick** [comment]
@jordanh any thoughts on the way forward? I definitely like the idea of a `History` table instead of just swapping out entire entities.
I also like the idea of storing the entire object instead of just a diff. The difference is gonna be 2-3 redundant fields. space is cheap.
I'm not sure if we need to store a `meetingId` to quickly show that the change happened in a meeting. What's the difference between a change happening in a meeting vs. 5 mins before? Not much, IMO.

**Comment by jordanh** [comment]
So glad you've started to think through this in detail.

Space _is_ cheap. I think that's the prevailing strategy. Here's what I think, building/adopting bits from the above:

- A table called `History`
- Store the entire object, not just a diff
- Store _every_ change, no matter if they occurred within 1 second of each other. Filter on report.
- Store the `meetingId`, under the same argument 'space is cheap.' Index it. Think of it as relating a bunch of objects to a changeset, rather than relying strictly on time.
- When changes occur outside of a meeting, say via a dashboard, `meetingId === null`
   - In the future, we could consider a field `changedUsing = enum('meeting', 'teamDash', 'meDash')` etc.

**Comment by mattkrick** [comment]
@jordanh this is really good, i like the enum!
One addtional thought: what about a `predecessor` field?
For example, let's say during a meeting a certain task changed ownership from Jordan to Matt. In the summary we wanna say something like "Jordan gave 'Make website' to Matt".
Without knowing the previous & current state, we'd have to do an `orderBy(r.desc('updatedBy')).skip(1)`. Not sure if it's premature optimization, but it seems like that query has the potential to get hot.

**Comment by jordanh** [comment]
Yeah, I like the `predecessor` field. I can see us using it.

**Comment by mattkrick** [comment]
one thing i forgot to mention about it... if we only store an ID on the predecessor field, then subscriptions will be difficult/impossible. But I can't imagine we'd ever need to subscribe on this stuff, right?

**Comment by jordanh** [comment]
I can't imagine. I think it'll be a paged interface...

**Comment by mattkrick** [comment]
I think we can get away with not creating a separate `Outcome` table, so we only have `Task` and `TaskHistory` (we can rename them later).

- `r.table('TaskHistory').getAll(taskId).orderBy(r.desc('updatedAt')).limit(1)[0]` must be a superset of `r.table('Task').get(taskId)`.
- The above is required primarily for Actions, because I am not permitted to see actions owned by anyone except me, so if I looked in the `Task` table, I might see the status of an action that was not initial. _However_, I can see Actions for any team member when that action was created in a meeting: `r.table('TaskHistory').getAll(taskId).orderBy('updatedAt').limit(1)[0]`.
- So, that means we can update tasks directly on the Task table while we're in a meeting. But when we do, we don't update the `MeetingHistory` _until_ the meeting comes to a close.  Then, we simply save every Task to TaskHistory, compute the diff, and turn it into human readable text.

**Comment by jordanh** [comment]
I just followed all that. LGTM! 👍

**Comment by mattkrick** [comment]
this looks pretty dern great!
Now we need to think about how to debounce the writes to ProjectHistory.
For example, if i change the owner & then change the content real quick, who cares? why do we need 2 records to show that?
This becomes even more critical as add per-keystroke propagation.
Since each write goes to a socket, that means 1 user is always going to use the same server. Therefore, think we can just put a debounce function on write to the ProjectHistory?

**Comment by jordanh** [comment]
Hmm. I think what we'd probably want to do is do a query for a document in `ProjectHistory` for a document with an `updatedAt > now() - DEBOUNCE_TIME`, and then upsert.

The upsert pattern with rethink can be found here: http://stackoverflow.com/questions/21109043/rethinkdb-how-to-do-real-upsert-insert-new-document-on-update-fail

So, sketching this out:

```javascript
r.table('ProjectHistory')
  .between(`now() - DEBOUNCE_TIME`, now(), { rightBound: 'closed', index: 'updatedAt' })
  .replace( (row) =>
    r.branch(
      row.eq(undefined),
      r.table('ProjectHistory').insert(`NEW_PROJECT_HISTORY_OBJECT`),
      row.merge(UPDATED_PROJECT_HISTORY_OBJECT)
    )
  )
```

**Comment by mattkrick** [comment]
@jordanh oh that's _good_! i was hoping there was a clever way to do exactly that!

Oh RethinkDB, you were so ahead of your time....

**Comment by mattkrick** [comment]
@jordanh Giving this some more thought, I think we should persist team name in the meeting history, too. Here's why:
- We can do away with querying the Team table
- I've got a sneaking suspicion that folks will start a team `Phase 1: Research` and then just rename that team (probably to save money) like `Phase 2: Prototype`. I want the history to show the Team name as it was when the meeting was had.

**Comment by mattkrick** [comment]
@jordanh giving this more thought, the algorithm isn't really a debounce, since it goes from the last write, not the next write.

For example:
- debouce in 5 seconds
- 0:00 i create a card
- 0:10 i add content, it puts a row in the DB
- 0:11 i change team member, it doesn't write a new row

Now if i look at the history, it isn't accurate, since that changed team member occurred too soon after the content change.

**Comment by jordanh** [comment]
Closed (for now) by #387
